01_fib/ailyntang/fib_test.go

```diff
@@ -0,0 +1,78 @@
+package main
+
+import (
+	"bytes"
+	"strconv"
+	"testing"
+)
+
+func TestMainOutput(t *testing.T) {
+	var buf bytes.Buffer
+	out = &buf
+
+	main()
+
+	expected := strconv.Quote("1\n1\n2\n3\n5\n8\n13\n")
+	actual := strconv.Quote(buf.String())
+
+	if expected != actual {
+		t.Errorf("Unexpected output from main()")
+		t.Errorf("expected: %v\nactual: %v", expected, actual)
+	}
+}
+
+func TestUnexpectedInputs(t *testing.T) {
+	errorString := strconv.Quote("fib.go only handles integers between 1 to 92 inclusive\n")
+
+	tests := []struct {
+		input    int
+		expected string
+	}{
+		{0, errorString},
+		{93, errorString},
+		{183, errorString},
+		{-5, errorString},
+		{-205, errorString},
+	}
+
+	for _, test := range tests {
+		var buf bytes.Buffer
+		out = &buf
+
+		fib(int64(test.input))
+
+		expected := test.expected
+		actual := strconv.Quote(buf.String())
+
+		if expected != actual {
+			t.Errorf("Unexpected response from fib(%v)", test.input)
+			t.Errorf("expected: %v\nactual: %v", expected, actual)
+		}
+	}
+}
+
+func TestFibonacciSequence(t *testing.T) {
+	tests := []struct {
+		input    int
+		expected string
+	}{
+		{1, strconv.Quote("1\n")},
+		{7, strconv.Quote("1\n1\n2\n3\n5\n8\n13\n")},
+		{10, strconv.Quote("1\n1\n2\n3\n5\n8\n13\n21\n34\n55\n")},
+	}
```

Comment:
Ideally you'd add a test for `fib(92)` because that is a boundary condition, but that makes for a very long string. Since this is the first lab I'll let it slide, but it's a good idea to always have tests for boundary conditions.

One way to make that easier to test would be to split up `fib` into `genFib()` which just returns the slice, and `fib()` calls `genFib()` and prints the result. That way the test here could just be testing `genFib()` and test that the length of the result is correct (92) and that the last element is correct.

As you work through the labs, have a think about how you might structure your code to make it more testable - you'll learn more of the Go language constructs so you'll have more options structuring your code (although you're already quite advanced in this code for the first lab :+1:)

Reply:
Great call out about testing boundaries and writing code to enable this. I'll keep that in mind for the future. 

Thanks for the tips on how to structure the test without typing out all 92 numbers. Very helpful.